internal/mtls/mtls.go

```diff
@@ -0,0 +1,130 @@
+package mtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/credentials"
+	exp_credentials "google.golang.org/grpc/experimental/credentials"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	hubbleRelayExpectedServerName = "ui.hubble-relay.cilium.io"
+)
+
+var (
+	ErrCertDataMissingInSecret = errors.New("required certificate data (ca.crt, tls.crt, or tls.key) not found in secret")
+)
+
+// getHubbleMTLSCertificatesFromSecret fetches ca.crt, tls.crt (client cert), and tls.key (client key)
+// from the specified Kubernetes secret.
+func getHubbleMTLSCertificatesFromSecret(
+	ctx context.Context,
+	clientset kubernetes.Interface,
+	logger *zap.Logger,
+	secretName string,
+	secretNamespace string,
+) (caData, clientCertData, clientKeyData []byte, err error) {
+
+	logger.Info("Fetching mTLS certificates from Kubernetes secret",
+		zap.String("secretName", secretName),
+		zap.String("secretNamespace", secretNamespace))
+
+	secret, err := clientset.CoreV1().Secrets(secretNamespace).Get(ctx, secretName, metav1.GetOptions{})
+	if err != nil {
+		logger.Error("Failed to get mTLS secret from Kubernetes API", zap.Error(err))
+		return nil, nil, nil, fmt.Errorf("failed to get secret '%s' in namespace '%s': %w", secretName, secretNamespace, err)
```

Comment:
No need to double-log.
Let the caller log this.
```suggestion
		return nil, nil, nil, fmt.Errorf("failed to get secret '%s' in namespace '%s': %w", secretName, secretNamespace, err)
```